modules/common/src/main/scala/surge/internal/health/windows/stream/sliding/SlidingHealthSignalStream.scala

```diff
@@ -77,31 +88,30 @@ private class SlidingHealthSignalStreamImpl(
     this
   }
 
-  override def processWindows(signalPatternMatchers: Seq[SignalPatternMatcher], monitoringActor: Option[ActorRef]): StreamHandle = {
-    val windowActors: Seq[HealthSignalWindowActorRef] = {
-      val advancerConfig = windowingConfig.advancerConfig
-      windowingConfig.frequencies
-        .map(freq => HealthSignalWindowActor(actorSystem, freq, WindowSlider(advancerConfig.advanceAmount, advancerConfig.buffer)))
-        .map(ref => ref.start(monitoringActor))
+  final override protected[health] def processWindows(): StreamHandle = {
+    if (Option(windowHandle).exists(h => h.isRunning)) {
+      throw new RuntimeException(
+        "SlidingHealthSignalStream is already processing signals." +
+          "Be sure to close the stream prior to processing again.")
     }
 
-    // todo: consider using an akka source
-    val future = Executors
-      .newCachedThreadPool()
-      .submit(new Callable[Unit] {
-        override def call: Unit = {
-          while (true) {
-            val signal = signals.take()
-            windowActors.foreach(a => a.processSignal(signal))
-          }
-        }
-      })
+    // Create a windowActor for each configured windowing frequency
+    val windowActors: Seq[HealthSignalWindowActorRef] = createWindowActors()
+
+    signalData = Source
+      .queue[HealthSignal](windowingConfig.maxStreamSize, OverflowStrategy.backpressure)
+      .toMat(Sink.foreach(signal => {
+        windowActors.foreach(actor => actor.processSignal(signal))
+      }))(Keep.left)
+      .run()(Materializer(actorSystem))
 
     // Stream Handle
-    () => {
-      future.cancel(true)
-      windowActors.foreach(actor => actor.stop())
-    }
+    new SlidingHealthSignalStreamHandle(
+      windowActors,
+      () => {
+        signalData = null
```

Comment:
I'd avoid this null - if it's something that may or may not exist make it an explicit `Option` so consumers are forced to deal with the case it's not there. If it exists as a non-optional type coming from scala code it's generally assumed that it's not null.

Reply:
yeah. this whole var vs val thing is a bit disconcerting.  Especially when the style guide always `var shames` me.  lol

Reply:
done